### PR TITLE
Use RWMutex instead of Mutex for locking templates

### PR DESCRIPTION
**Pull request by zeripath**
The template render lock means that template rendering in unrolled is single-threaded
and only one template can be rendered at a time by any project using unrolled. This is
clearly not ideal.

This PR suggests using a RWMutex instead which restores concurrent rendering to unrolled.

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

**Comment by zeripath** [comment]
I have an additional pr that can use fsnotify to recompile the templates on change to their directories.

And for the asset routes the compileTemplates function should just be exposed so that a user like gitea can decide not to use isDevelopment but use their own function to call for recompilation

**Comment by unrolled** [comment]
This is an awesome find, thanks for the performance boost 😄 

I checked in a quick benchmark to test the difference... your fix definitely helps a lot:
```bash
goos: darwin
goarch: amd64
pkg: github.com/unrolled/render
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
(Before) BenchmarkHTML-12    	  500048	      2170 ns/op	    1386 B/op	      20 allocs/op
(After)  BenchmarkHTML-12    	 1909053	     623.0 ns/op	    1387 B/op	      20 allocs/op
change
PASS
ok  	github.com/unrolled/render	2.366s
```

Looking forward to the other PRs 👍

File: render.go [hunk 1]
```diff
@@ -389,14 +397,15 @@ func (r *Render) Data(w io.Writer, status int, v []byte) error {
...
 func (r *Render) HTML(w io.Writer, status int, name string, binding interface{}, htmlOpt ...HTMLOptions) error {
-	r.templatesLk.Lock()
-	defer r.templatesLk.Unlock()
 
 	// If we are in development mode, recompile the templates on every HTML request.
 	if r.opt.IsDevelopment {
 		r.compileTemplates()
 	}
 
+	r.templatesLk.RLock()
```
Comment (lunny) [inline render.go:406]: Is this lock/unlock necessary? Since all other locks just in compileTemplates ?

Comment (zeripath) [reply render.go:406]: Yes because this is using the templates. We can't replace the templates until they're not being used by anyone else.

**Pull request merged**